### PR TITLE
feat(TableRouteValue): add panic notes and type checks

### PR DESCRIPTION
**Pull request by AntiTopQuark**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

add panic notes and type checks for TableRouteValue

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
close https://github.com/GreptimeTeam/greptimedb/issues/3016

### CODE REVIEW

**Comment by AntiTopQuark** [comment]
Previous PR:
https://github.com/GreptimeTeam/greptimedb/pull/3023

File: src/common/meta/src/key/table_route.rs [hunk 1]
```diff
@@ -62,29 +62,57 @@ impl TableRouteValue {
...
-        })
+        }))
     }
 
     /// Returns the version.
     ///
     /// For test purpose.
+    ///
+    /// # Panics
+    /// The route type is not the [TableRouteValue::Physical].
```
Comment (WenyXu) [inline src/common/meta/src/key/table_route.rs:87]:
```suggestion
```

File: src/common/meta/src/key/table_route.rs [hunk 2]
```diff
@@ -62,29 +62,57 @@ impl TableRouteValue {
...
+        );
+        Ok(self.physical_table_route().version)
     }
 
     /// Returns the corresponding [RegionRoute].
-    pub fn region_route(&self, region_id: RegionId) -> Option<RegionRoute> {
-        self.physical_table_route()
+    ///
+    /// # Panics
+    /// The route type is not the [TableRouteValue::Physical].
```
Comment (WenyXu) [inline src/common/meta/src/key/table_route.rs:102]:
```suggestion
```

File: src/common/meta/src/key/table_route.rs [hunk 3]
```diff
@@ -96,8 +124,14 @@ impl TableRouteValue {
     ///
     /// # Panics
     /// The route type is not the [TableRouteValue::Physical].
```
Comment (WenyXu) [inline src/common/meta/src/key/table_route.rs:126]:
```suggestion
```

File: src/common/meta/src/error.rs [hunk 4]
```diff
@@ -330,6 +330,9 @@ pub enum Error {
 
     #[snafu(display("The topic pool is empty"))]
     EmptyTopicPool { location: Location },
+
+    #[snafu(display("Unexpected table route type: {}", err_msg))]
+    UnexpectedTableRouteType { location: Location, err_msg: String },
```
Comment (WenyXu) [inline src/common/meta/src/error.rs:335]: I prefer to use a `UnexpectedLogicalRouteTable`.

File: src/common/meta/src/key/table_route.rs [hunk 5]
```diff
@@ -62,29 +62,57 @@ impl TableRouteValue {
     }
 
     /// Returns a new version [TableRouteValue] with `region_routes`.
-    pub fn update(&self, region_routes: Vec<RegionRoute>) -> Self {
+    ///
+    /// # Panics
+    /// The route type is not the [TableRouteValue::Physical].
```
Comment (MichaelScofield) [inline src/common/meta/src/key/table_route.rs:67]: Will not panic now:
```suggestion
    /// Returns [UnexpectedTableRouteType] if the route type is not the [TableRouteValue::Physical].
```
same for other places below

File: src/meta-srv/src/selector/load_based.rs [hunk 6]
```diff
@@ -143,10 +143,14 @@ async fn get_leader_peer_ids(
...
         .map(|route| {
             route.map_or_else(Vec::new, |route| {
-                find_leaders(route.region_routes())
-                    .into_iter()
-                    .map(|peer| peer.id)
-                    .collect()
+                find_leaders(
+                    route
+                        .region_routes()
+                        .expect("expected physical table route"),
```
Comment (MichaelScofield) [inline src/meta-srv/src/selector/load_based.rs:149]: The "expect" is not safe here. You can `ensure!` the table route is of type physical above first.

File: src/meta-srv/src/selector/load_based.rs [hunk 7]
```diff
@@ -143,10 +143,14 @@ async fn get_leader_peer_ids(
...
+                )
+                .into_iter()
+                .map(|peer| peer.id)
+                .collect()
```
Comment (WenyXu) [inline src/meta-srv/src/selector/load_based.rs:153]: Try this
```rust
    table_metadata_manager
        .table_route_manager()
        .get(table_id)
        .await
        .context(error::TableMetadataManagerSnafu)
        .map(|route| {
            route.map_or_else(
                || Ok(Vec::new()),
                |route| {
                    let region_routes = route
                        .region_routes()
                        .context(error::UnexpectedTableRouteTypeSnafu { err_msg: "" })?;

                    Ok(find_leaders(region_routes)
                        .into_iter()
                        .map(|peer| peer.id)
                        .collect())
                },
            )
        })?
```

**Review by WenyXu — APPROVED** [review]
LGTM

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**